server/core/src/main/java/org/infinispan/server/core/RequestTracer.java

```diff
@@ -0,0 +1,97 @@
+package org.infinispan.server.core;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+
+import org.infinispan.commons.logging.Log;
+import org.infinispan.commons.logging.LogFactory;
+
+import io.opentracing.Span;
+import io.opentracing.Tracer;
+
+public class RequestTracer {
+   private static final Log log = LogFactory.getLog(RequestTracer.class);
+   private static TracerAdapter adapter;
+
+   public static final String TRACER_FACTORY_CLASS = System.getProperty("infinispan.opentracing.factory.class");
+
+   public static final String TRACER_FACTORY_METHOD = System.getProperty("infinispan.opentracing.factory.method");
+
+   static {
+      if (TRACER_FACTORY_CLASS != null && TRACER_FACTORY_METHOD != null) {
+         try {
+            Class<?> tracerFactoryClass =
+                  Thread.currentThread().getContextClassLoader().loadClass(RequestTracer.TRACER_FACTORY_CLASS);
+            Method getTracer = tracerFactoryClass.getMethod(TRACER_FACTORY_METHOD);
+            Object factoryInstance;
+            if (Modifier.isStatic(getTracer.getModifiers())) {
+               factoryInstance = null;
+            } else {
+               factoryInstance = tracerFactoryClass.getDeclaredConstructor().newInstance();
+            }
+            adapter = new TracerAdapter((Tracer) getTracer.invoke(factoryInstance));
+            log.infof("OpenTracing implementation loaded: %s", adapter);
+         } catch (Throwable e) {
+            log.warnf(e, "OpenTracing implementation could not be loaded");
+         }
+      } else {
+         log.info("OpenTracing integration is disabled");
+      }
+   }
+
+   public static void start() {
+      // init() method call is only needed to trigger the static initializer
+   }
+
+   public static void stop() {
+      if (adapter != null) {
+         adapter.stop();
+      }
+   }
+
+   public static Object requestStart(String operationName) {
```

Comment:
Why aren't we returning `io.opentracing.Span` here?

We're using the opentracing api to provide implementation independence, so I don't think it's an issue to expose this interface outside of this class. It would make the consumer classes more explicit if they didn't have to pass around instances of `Object`.

Reply:
The return value is `Object` because I developed this first as a patch for the WildFly-based server (https://github.com/infinispan/jdg/pull/1647).

Because API versions 0.31.0 and 0.33.0 are incompatible, putting the API on the boot classpath with the implementation allows us to support both the AppDynamics tracer (which only works with 0.31.0), and the latest Jaeger client (which requires 0.33.0) with the same server. But the server must be able to run without any opentracing-api jar, so I had to make the return type `Object`.

I was hoping in master I could put the implementation jars in `server/lib` instead of modifying the boot classpath, but they're not quite equivalent: `RequestTracer` is in the application classloader, which doesn't have access to the `server/lib` classes.

In this initial PR I made a compromise: I put opentracing-api version 0.31.0 in the server's `/lib` directory. This means an impl that uses 0.31.0 can be just dropped in `/server/lib`, which made the integration test easier to write, but running with the latest Jaeger client requires overwriting the opentracing-api jar in `/lib`.

For the future, I'm considering several options:
* Invoke all opentracing methods via reflection, using the extensions classloader to load interfaces. This would allow everything to stay in `/server/lib`, including opentracing-api.
* Do not include the opentracing-api in `/lib`, but require all the implementation jars to be copied to `/lib`, not `/server/lib`, and extend the server test drivers to make it possible.
* Include a default impl (the latest Jaeger client) in `/lib/jaeger`, but allow users to unpack another impl in `/lib/othertracingimpl` and select it with `-Dopentracing.impl.directory=othertracingimpl`.

Reply:
Thanks for the detailed explanation :+1: